### ISSUE TITLE
how to avoid high RAM usage

### ISSUE THREAD

**Issue by spasserkongen** [ultralytics/ultralytics#1467]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am trying to train yolov8 on a custom dataset (>4M images, >20M BBs) and the RAM usage is extremely high, exausting my 128GB.
I launch training like this:
    results = model.train(data="data.yaml", project="view", name="yolov8", batch=5, epochs=3, workers=15, mosaic=0.8,
                          plots=True, resume=False, degrees=0.15, save=True, patience=2, imgsz=640, close_mosaic=1, v5loader=False)

What can I do, are there any options to not keep everything in memory (or whatever takes up the entire RAM)?

**Comment by glenn-jocher** [comment]
@spasserkongen if you can provide a way for us to download your dataset and reproduce your RAM usage issue then we might be able to help. Other than that use a smaller dataset or a machine with more RAM.


## How to create a Minimal, Reproducible Example

When asking a question, people will be better able to provide help if you provide **code** that they can easily understand and use to **reproduce** the problem. This is referred to by community members as creating a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Your code that reproduces the problem should be:

- ✅ **Minimal** – Use as little code as possible to produce the problem
- ✅ **Complete** – Provide all parts someone else needs to reproduce the problem
- ✅ **Reproducible** – Test the code you're about to provide to make sure it reproduces the problem

For [Ultralytics](https://ultralytics.com/) to provide assistance your code should also be:

- ✅ **Current** – Verify that your code is up-to-date with GitHub with `git pull` or `pip install -U ultralytics` to ensure your problem has not already been solved in the latest code version.
- ✅ **Unmodified** – Your problem must be reproducible using official Ultrlaytics code without changes. [Ultralytics](https://ultralytics.com/) does not provide support for custom code ⚠️.

If you believe your problem meets all the above criteria, please close this issue and raise a new one using the 🐛 **Bug Report** [template](https://github.com/ultralytics/ultralytics/issues/new/choose) with a [minimum reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) to help us better understand and diagnose your problem. 

Thank you! 😃

**Comment by spasserkongen** [comment]
There is no feasible way for me to supply you the dataset.

Is there no support for the framwork to load data on-the-go from harddrive? Like e.g. Darknet? How do other handle large datasets?

**Comment by glenn-jocher** [comment]
@spasserkongen yes, data is loaded only when it is needed. If you pass the optional `cache` argument then it is cached on disk or in RAM.

**Comment by bnascimento** [comment]
hi guys, I dont want to interrupt or hijack the thread in anyform, but I think both @spasserkongen and myself have the same problem. see https://github.com/ultralytics/ultralytics/issues/617 
we suspect this is due to some memory leak, can you check if you use close_mosaic with the same number of epochs that you train on (in order to prevent mosaic augmentation to take place) if that aliviates some of the memory consumption?

**Comment by Laughing-q** [comment]
@spasserkongen reduce `worker` might help.

**Comment by achbogga** [comment]
I have the same problem. With large datasets, the RAM is being accumulated while training! I suspect a memory leak.

**Comment by glenn-jocher** [comment]
@achbogga thank you for bringing this to our attention. We apologize for any inconvenience this memory leak issue has caused. 

Our team is currently investigating to identify the source of the problem and will work to resolve it as soon as possible. One potential workaround for now could be to reduce the batch size or image size, which may help alleviate the issue. 

If you have any additional information or insights that may help us resolve this problem, please do not hesitate to share them. Thank you for your patience and understanding.

**Comment by achbogga** [comment]
@glenn-jocher Just to give you more insights into reproducing this issue: if you have more than 1000 masks per image, you will most likely run into this issue. If your labels are well below this number, you might not be able to reproduce this issue. Also, I was running this on multiple GPUs. This error won't occur if you are using a single GPU for this massive dataset. The RAM increase also happens only for the first epoch and stays relatively flat during the training! Hope this helps!

**Comment by glenn-jocher** [comment]
@achbogga thank you for sharing your experience with running YOLOv8 on large datasets. Based on your insights, it appears that this memory leak issue may occur when there are more than 1000 masks per image, and may be more pronounced when running on multiple GPUs. Additionally, the RAM increase seems to happen only during the first epoch and remains relatively stable thereafter. We hope this information is helpful in reproducing and addressing the issue. Please let us know if you have any further questions or concerns.

**Comment by pavelshutov-pr3vision** [comment]
Hello! 

I got the similar problem with the large dataset (about 1.7M images) and found that the reason may be related with this [pytorch issue](https://github.com/pytorch/pytorch/issues/13246). I noticed that `labels`, `ims`, `im_files`, `npy_files` in `BaseDataset` stored in python list and in accordance with the given pytorch issue it may cause this strange behavior when `workers > 0`.

So I tried converting labels, ims, im_files, npy_files to bytearrays and np.arrays (by a very dumb way) and this modification reduced the increase of memory usage during dataloader iterations (from tens of GB to a few GB in my case). I would mention that converting to torch.tensor in my experiments also solved the problem, but caused an extreme increase of memory usage after about 5000 iterations.

In the following graph, I compare memory usage with and without modifications (for a detection dataset with approximately 1.7M images, batch_size=64 and workers=18):

![image](https://github.com/ultralytics/ultralytics/assets/115560499/590e128b-57c1-4c00-a8dd-427247e4ca05)

I will briefly describe the modification for the `bytearray + np.array` case, it may also be useful:

* At the end of `BaseDataset.__init__` I added this conversion code (really slow code):
```
self.labels_data_keys = []
self.labels_data, self.labels_data_offsets, self.labels_data_lengths = [], [], []
self.images_data, self.images_data_offsets, self.images_data_lengths = [], [], []

for label in self.labels:
    self.labels_data_keys += [label_key for label_key in label.keys() if label_key not in self.labels_data_keys]
    label_list = [label[key] if key in label else None for key in self.labels_data_keys]
    encoded_data = list(pickle.dumps(label_list))
    self.labels_data_offsets.append(len(self.labels_data))
    self.labels_data_lengths.append(len(encoded_data))
    self.labels_data += encoded_data

for i in range(len(self.ims)):
    data = [self.ims[i], self.im_files[i], self.npy_files[i]]
    encoded_data = list(pickle.dumps(data))
    self.images_data_offsets.append(len(self.images_data))
    self.images_data_lengths.append(len(encoded_data))
    self.images_data += encoded_data

self.labels_data_keys = bytearray(pickle.dumps(self.labels_data_keys))
self.labels_data = bytearray(self.labels_data)
self.labels_data_offsets = np.array(self.labels_data_offsets, dtype=np.int32)
self.labels_data_lengths = np.array(self.labels_data_lengths, dtype=np.int32)
self.images_data = bytearray(self.images_data)
self.images_data_offsets = np.array(self.images_data_offsets, dtype=np.int32)
self.images_data_lengths = np.array(self.images_data_lengths, dtype=np.int32)
```
* In `BaseDataset.load_image` I made the following modification:
```
# im, f, fn = self.ims[i], self.im_files[i], self.npy_files[i]
offset, length = self.images_data_offsets[i], self.images_data_lengths[i]
im, f, fn = pickle.loads(bytes(self.images_data[offset: offset + length]))
```
* And in `BaseDataset.get_label_info` I did the following:
```
#label = deepcopy(self.labels[index])  # requires deepcopy() https://github.com/ultralytics/ultralytics/pull/1948
offset, length = self.labels_data_offsets[index], self.labels_data_lengths[index]
data = pickle.loads(bytes(self.labels_data[offset: offset + length]))
label = {pickle.loads(bytes(self.labels_data_keys))[i]: d for i, d in enumerate(data)}
```

I think that for the data types used in `BaseDataset`/`YOLODataset`, a more convenient and powerful representation can be found that would not cause this strange behavior using RAM.

If it's helpful, I think I can also provide an example of minimal code to reproduce with synthetic data that will cause issues with increased RAM usage.

**Comment by glenn-jocher** [comment]
@pavelshutov-pr3vision hello,

Thank you for sharing your findings regarding the memory leak issue you encountered while running the YOLOv8 model on a large dataset with a high number of workers. It appears that the issue may be related to the pytorch issue you mentioned, which can cause unexpected behavior when `workers > 0`. 

Based on your experiments, you found that converting `labels`, `ims`, `im_files` and `npy_files` to `bytearrays` and `np.arrays` helped reduce the RAM usage during dataloader iterations. However, you also found that converting to `torch.tensor` caused an extreme increase in memory usage after a certain number of iterations.

Thank you for providing the modification code you used to convert the data to `bytearrays` and `np.arrays`. Your insights may be helpful to others who encounter a similar issue while running the YOLOv8 model on large datasets. 

If you have a minimal code example to reproduce this issue with synthetic data, please feel free to share it as well.

Thank you for your contribution to the YOLOv8 community.

**Comment by pavelshutov-pr3vision** [comment]
@glenn-jocher I wrote a small synthetic example where the current problem is clearly visible. I presented this as a small benchmark that will show the increase of memory consumption before and after the modification. I tried this code on a Linux system, but on Windows it consumes too much memory just to initialize the data loader, even with the reduced number of worker processes. So I couldn't do a normal test on Windows.

Example code:
```
import pickle
import psutil
from copy import deepcopy

import tqdm
import matplotlib.pyplot as plt
from matplotlib.ticker import AutoMinorLocator, MultipleLocator
import numpy as np

from ultralytics.yolo.data import build_dataloader
from ultralytics.yolo.data.dataset import YOLODataset


class SyntheticHyp:
    overlap_mask, mask_ratio, hsv_h, hsv_s, hsv_v = True, 4, 0.015, 0.7, 0.4
    degrees, translate, scale, shear, perspective = 0.0, 0.1, 0.5, 0.0, 0.0
    flipud, fliplr, mosaic, mixup, copy_paste = 0.0, 0.5, 1.0, 0.0, 0.0


class SyntheticDataset(YOLODataset):
    def synthetic_data_init(self, size: int, batch_size: int = 64, hyp=None):
        nc, nb = 10, 3
        self.augment, self.imgsz, self.batch_size = True, 320, batch_size
        self.ni, self.pad, self.rect, self.single_cls = size, 0.0, False, False
        self.stride, self.use_keypoints, self.use_segments = 32, False, False

        self.data = {
            "path": "", "train": "", "val": "", "test": "",
            "names": {i: f"class{i}" for i in range(nc)}, "yaml_file": "", "nc": nc,
        }
        synthetic_label = {
            "im_file": f"/just/random/path/not/used/temp.temp", "shape": (1024, 1024),
            "cls": np.random.randint(nc, size=[nb, 1]).astype(np.float32),
            "bboxes": np.zeros([nb, 4], dtype=np.float32),
            "segments": [], "keypoints": None, "normalized": True, "bbox_format": "xywh",
        }
        synthetic_label["bboxes"][:, 2:] = 1.0
        self.labels = [deepcopy(synthetic_label) for _ in range(self.ni)]
        self.transforms = self.build_transforms(hyp)
        self.modified = False

    def modify(self):
        # Here only labels are converted due to disabled self.load_image in this example
        # (so there is no need to convert ims, im_files and npy_files here)
        self.labels_data_keys = []
        self.labels_data, self.labels_data_offsets, self.labels_data_lengths = [], [], []

        for label in self.labels:
            self.labels_data_keys += [
                label_key for label_key in label.keys() if label_key not in self.labels_data_keys
            ]
            label_list = [label[key] if key in label else None for key in self.labels_data_keys]
            encoded_data = list(pickle.dumps(label_list))
            self.labels_data_offsets.append(len(self.labels_data))
            self.labels_data_lengths.append(len(encoded_data))
            self.labels_data += encoded_data

        self.labels_data_keys = bytearray(pickle.dumps(self.labels_data_keys))
        self.labels_data = bytearray(self.labels_data)
        self.labels_data_offsets = np.array(self.labels_data_offsets, dtype=np.int32)
        self.labels_data_lengths = np.array(self.labels_data_lengths, dtype=np.int32)
        self.modified = True

    def get_label_info(self, index):
        """Get and return label information from the dataset."""
        if self.modified:
            offset, length = self.labels_data_offsets[index], self.labels_data_lengths[index]
            data = pickle.loads(bytes(self.labels_data[offset : offset + length]))
            label = {
                pickle.loads(bytes(self.labels_data_keys))[i]: d for i, d in enumerate(data)
            }
        else:
            label = deepcopy(self.labels[index])

        shape = label["shape"]
        label.pop("shape", None)  # shape is for rect, remove it
        # label['img'], label['ori_shape'], label['resized_shape'] = self.load_image(index)
        label["img"], label["ori_shape"], label["resized_shape"] = (
            np.zeros((320, 320, 3), dtype=np.uint8), shape, (320, 320)
        )

        label["ratio_pad"] = (
            label["resized_shape"][0] / label["ori_shape"][0],
            label["resized_shape"][1] / label["ori_shape"][1],
        )  # for evaluation
        if self.rect:
            label["rect_shape"] = self.batch_shapes[self.batch[index]]
        label = self.update_labels_info(label)
        return label


if __name__ == "__main__":
    synthetic_dataset_size = 2000000
    batch_size, num_workers = 64, 12
    iterations_limit = 5000

    dataset = SyntheticDataset.__new__(SyntheticDataset)
    dataset.synthetic_data_init(synthetic_dataset_size, batch_size, SyntheticHyp)

    memory_consumption = {"default": [], "modified": []}

    # Run a memory consumption benchmark:
    for test_name in ["default", "modified"]:
        if test_name == "modified": dataset.modify()
        dataloader = build_dataloader(dataset, batch_size, num_workers, True, -1)

        for i, batch in enumerate(
            tqdm.tqdm(dataloader, total=iterations_limit, desc=f"{test_name} test")
        ):
            total_used_memory = psutil.virtual_memory().used
            memory_consumption[test_name].append(round(total_used_memory / (1024**3), 3))
            if i + 1 >= iterations_limit: break

        del dataloader
        memory_consumption[test_name] = np.array(memory_consumption[test_name], dtype=float)
        memory_consumption[test_name] -= memory_consumption[test_name][0]

    # Plotting results:
    fig, ax = plt.subplots(1, 1, figsize=(6, 4))
    ax.plot(memory_consumption["default"], color="#ed4c84", label="Default")
    ax.plot(memory_consumption["modified"], color="#4cb7ed", label="bytearray + np.array")

    ax.set_xlabel("Iteration", fontsize=16)
    ax.set_ylabel("Increasing RAM usage (GB)", fontsize=16)

    ax.yaxis.set_minor_locator(AutoMinorLocator(2))
    ax.yaxis.set_major_locator(MultipleLocator(2))

    ax.grid(which="major", alpha=0.1)
    ax.grid(which="minor", alpha=0.05)
    ax.legend(loc="upper left")
    plt.tight_layout()
    plt.show()
```

During running this on Linux (python 3.10.6, pytorch 1.13.1, ultralytics 8.0.99), I got the following plot:
![2](https://github.com/ultralytics/ultralytics/assets/115560499/b7bc082a-138b-409d-a405-ed9ed21d59b9)

**Comment by glenn-jocher** [comment]
@pavelshutov-pr3vision got it. Your results look good! Can you please submit a PR with your proposed updates? Then I can test the PR branch. Thanks!

**Comment by pavelshutov-pr3vision** [comment]
@glenn-jocher I haven't given enough time yet to prepare the PR and after #2791 I found that it became a lot more difficult to come up with a Python List replacement in the yolo dataset. As long as these lists did not change during the iteration of the dataset, they could be converted to ```bytearrays```. But now using buffering causes changes in the lists and I don't know how to get around this problem.

Lately, I've also found a nice [blog post](https://ppwwyyxx.com/blog/2022/Demystify-RAM-Usage-in-Multiprocess-DataLoader/) that explains the memory consumption issue in more details. This post also suggests serialization to ```torch.Tensor``` as a solution to avoid memory issues. But this serialization also does not involve mutation and is not suitable for use with buffering during data set iteration.

At the moment, it looks like only one thing should be used: either avoiding high memory usage with large datasets by serialization, or speeding up the dataset by buffering. But both of these things are very good for large datasets and I would like to find a way to solve the memory problem while maintaining buffering :slightly_smiling_face:

**Comment by glenn-jocher** [comment]
Thanks for the update, @pavelshutov-pr3vision. It sounds like the situation is a bit more complex now with the changes in the yolo dataset. I appreciate that you're still working on finding a solution that maintains buffering while avoiding high memory usage with large datasets. It's great that you came across the blog post that explains the memory consumption issue in more detail and suggests serialization to `torch.Tensor` as a solution. While it may involve avoiding mutation and being less suitable for use with buffering during dataset iteration, serialization to `torch.Tensor` might still be a viable option for reducing memory usage. I wish you the best of luck in continuing your exploration to find a solution that works for your needs.

**Comment by davidsvaughn** [comment]
This sounds a lot like issue [#10383](https://github.com/ultralytics/yolov5/pull/10383) that I (and others) had with yolov5.  I'll have to investigate to be sure, but if it is the same issue, there was a pretty easy fix that just involved partitioning the training images among the N available GPUs so that each DDP process cached only 1/N of the images into RAM.  The solution described above might be the "correct" way to fix this, but it looks very complicated, and seems to have gone stale.  If this _is_ the same issue as [#10383](https://github.com/ultralytics/yolov5/pull/10383) then I might try to replicate that fix here (for yolov8) and submit a PR....

**Comment by glenn-jocher** [comment]
@davidsvaughn thank you for the suggestion. We appreciate any proposed solutions that could potentially help resolve this ongoing issue. Partitioning the training images among the available GPUs would indeed be an interesting approach to alleviate the memory consumption issue. 

We would absolutely encourage you to try and adapt the fix to the YOLOv8 context. Feel free to share any findings you have, and if it proves successful, a new PR would definitely be welcome. Your contribution could potentially help a lot of users deal with similar issues. 

We're grateful for your valuable input and look forward to seeing your progress.

**Comment by babyta** [comment]
> @spasserkongen reduce `worker` might help.

useful,thank you!!

**Comment by glenn-jocher** [comment]
@babyta, I'm glad to hear that reducing the number of workers helped with your problem! That generally alleviates memory usage in the data loader component during training. Feel free to reach out if you encounter any more issues or have further questions. Keep up the good work with YOLOv8!

**Comment by westlinkin** [comment]
I have a similar issue. It happens when training the segmentation model. While executing 'Plotting labels to runs/segment/train/labels.jpg... ', it takes a crazy amount of RAM. ( training object detection model okay).

My dataset has less than 1000 images.

**Comment by glenn-jocher** [comment]
@westlinkin yes, setting `plots` to False will disable the generation of label plots during training, which can reduce RAM usage significantly, especially for segmentation tasks with large label sets. If you don't need to visualize the labels, this is a good workaround. 👍

**Comment by JulioZhao97** [comment]
When training on a dataset consisting of 3M images, the training crashed within one epoch

**Comment by babyta** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by glenn-jocher** [comment]
Hello @babyta,

Thank you for reaching out. It seems like you're encountering an issue with training on a large dataset of 3 million images, where the training crashes within one epoch. Let's work together to resolve this.

### Steps to Troubleshoot:

1. **Minimum Reproducible Example**:
   Could you please provide a minimum reproducible code example? This will help us understand the exact setup and parameters you're using. You can refer to our [Minimum Reproducible Example Guide](https://docs.ultralytics.com/help/minimum_reproducible_example) for more details on how to create one. This is crucial for us to reproduce the issue and investigate a solution.

2. **Package Versions**:
   Ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade your packages using the following commands:
   ```bash
   pip install --upgrade torch ultralytics
   ```
   After upgrading, please try running your training again to see if the issue persists.

3. **Memory Management**:
   Given the size of your dataset, memory management is critical. Here are a few tips that might help:
   - **Reduce Workers**: Lowering the number of workers can help manage memory usage better.
   - **Disable Plots**: As you mentioned earlier, setting `plots=False` can significantly reduce RAM usage.
   - **Batch Size**: Ensure your batch size is appropriate for your available GPU memory. You might need to reduce it if you're running out of memory.

4. **Segmentation-Specific Tips**:
   - **Image Size**: Ensure that the image sizes are not excessively large, as this can quickly consume memory.
   - **Data Augmentation**: Be cautious with heavy data augmentation techniques that can increase memory usage.

### Example Configuration:
Here's an example of how you might configure your training to be more memory-efficient:
```python
results = model.train(
    data="data.yaml",
    project="view",
    name="yolov8",
    batch=4,  # Adjust based on your GPU memory
    epochs=3,
    workers=4,  # Reduce number of workers
    mosaic=0.8,
    plots=False,  # Disable plots to save memory
    resume=False,
    degrees=0.15,
    save=True,
    patience=2,
    imgsz=640,
    close_mosaic=1,
    v5loader=False
)
```

Please try these suggestions and let us know if the issue persists. We're here to help!
